File: sdk/rust/src/artifact/go.rs
use crate::{
    api,
    api::artifact::ArtifactSystem::{
        Aarch64Darwin, Aarch64Linux, UnknownSystem, X8664Darwin, X8664Linux,
    },
    artifact::{step, system, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::{bail, Result};

/// Builds the shared source used by the Go tooling artifacts (`goimports`, `gopls`).
#[must_use]
pub fn source_tools(name: &str) -> api::artifact::ArtifactSource {
    let version = "0.42.0";

    let path = format!("https://sdk.vorpal.build/source/go-tools-v{version}.tar.gz");

    ArtifactSource::new(name, path.as_str()).build()
}

/// Build-target `go` toolchain, fetched as a prebuilt binary release.
#[derive(Default)]
pub struct Go {}

impl Go {
    /// Creates a builder for the `go` toolchain.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Builds the `go` artifact for the context's target system.
    ///
    /// # Errors
    ///
    /// Returns an error if the target system has no known `go` release, or if registering the
    /// source or artifact with the build context fails.
    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "go";

        let system = context.get_system();

        let source_target = match system {
            Aarch64Darwin => "darwin-arm64",
            Aarch64Linux => "linux-arm64",
            X8664Darwin => "darwin-amd64",
            X8664Linux => "linux-amd64",
            UnknownSystem => bail!("unsupported {name} system: {}", system.as_str_name()),
        };

        let source_version = "1.26.0";
        let source_path =
            format!("https://sdk.vorpal.build/source/go{source_version}.{source_target}.tar.gz");

        let source = ArtifactSource::new(name, source_path.as_str()).build();

        let step_script = format!("cp -pr \"./source/{name}/go/.\" \"$VORPAL_OUTPUT\"");
        let steps = vec![step::shell(context, &[], &[], step_script, &[]).await?];

        Artifact::new(name, steps, system::SYSTEMS)
            .with_aliases(vec![format!("{name}:{source_version}")])
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/go.ts
import type {
  ArtifactSource as ArtifactSourceMsg,
} from "../api/artifact/artifact.js";
import { ArtifactSystem } from "../api/artifact/artifact.js";
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";

/**
 * Creates a shared ArtifactSource for the Go tools repository.
 * Used by goimports and gopls which both build from the same source.
 *
 * Mirrors Rust `go::source_tools()` in `sdk/rust/src/artifact/go.rs`.
 */
export function sourceTools(name: string): ArtifactSourceMsg {
  const version = "0.42.0";
  const path = `https://sdk.vorpal.build/source/go-tools-v${version}.tar.gz`;
  return new ArtifactSource(name, path).build();
}

// ---------------------------------------------------------------------------
// Go Distribution
// ---------------------------------------------------------------------------

/**
 * Builder for the Go distribution artifact.
 *
 * Mirrors Rust `Go` struct in `sdk/rust/src/artifact/go.rs`.
 * Downloads and extracts the official Go binary distribution.
 */
export class GoBin {
  async build(context: ConfigContext): Promise<string> {
    const name = "go";
    const system = context.getSystem();

    let sourceTarget: string;

    switch (system) {
      case ArtifactSystem.AARCH64_DARWIN:
        sourceTarget = "darwin-arm64";
        break;
      case ArtifactSystem.AARCH64_LINUX:
        sourceTarget = "linux-arm64";
        break;
      case ArtifactSystem.X8664_DARWIN:
        sourceTarget = "darwin-amd64";
        break;
      case ArtifactSystem.X8664_LINUX:
        sourceTarget = "linux-amd64";
        break;
      default:
        throw new Error(`unsupported ${name} system: ${system}`);
    }

    const sourceVersion = "1.26.0";
    const sourcePath = `https://sdk.vorpal.build/source/go${sourceVersion}.${sourceTarget}.tar.gz`;

    const source = new ArtifactSource(name, sourcePath).build();

    const stepScript = `cp -pr "./source/${name}/go/." "$VORPAL_OUTPUT"`;
    const steps = [await shell(context, [], [], stepScript, [])];
    const systems = [
      ArtifactSystem.AARCH64_DARWIN,
      ArtifactSystem.AARCH64_LINUX,
      ArtifactSystem.X8664_DARWIN,
      ArtifactSystem.X8664_LINUX,
    ];

    return new Artifact(name, steps, systems)
      .withAliases([`${name}:${sourceVersion}`])
      .withSources([source])
      .build(context);
  }
}
